File: euler15.py
import operator
from itertools import izip
from pprint import pprint


def zip_with(op, r1, r2):
    return [op(*i) for i in izip(r1, r2)]


def pascal(row_limit):
    row = [1]
    rows = [row]
    for _ in range(0, row_limit):
        row = zip_with(operator.add, [0] + row, row + [0])
        rows.append(row)
    return rows

pprint(pascal(40)[40][20])

File: euler15.hs
-- Totally stumped on this one

-- This is brilliant from the haskell wiki
-- http://www.haskell.org/haskellwiki/Blow_your_mind

-- The initial row is [1], iterate the applies the row function which applies +
-- to all members for the two row lists then zips the lists together
pascal = iterate (\row -> zipWith (+) ([0] ++ row) (row ++ [0])) [1]

main = print $ pascal !! 40 !! 20